types/bun/.eslintrc.json

```diff
@@ -0,0 +1,11 @@
+{
+    "rules": {
```

Comment:
These are real lints you probably don't want to disable.

Reply:
All of these were disabled with intentionality. A lot of these rules are unnecessarily onerous for something with the API surface of a runtime like Bun. I saw that @types/node has disabled various eslint/tslint rules for similar reasons so I figured it was kosher. Happy to add additional context but I wasn't disabling rules willy-nilly.

Here's some context:

- `no-single-declare-module`: Doesn't apply at all to typings packages that declare types for multiple modules.
- `no-single-element-tuple-type`: Breaks for rest parameters: `function <T extends SomeTuple>stuff(...args: [...T]){}`. There are also lots of times where a single-element tuples are useful in conditional type logic.
- `no-unnecessary-generics`: This one really overreaches, any method like `<T>name(): Something<T>` triggers this. That's a common pattern for methods where passing a generic as part of the user-facing API is expected. 
- `redundant-undefined`: This doesn't factor in the existence of strictOptionalPropertyTypes...very bad. This should be disabled across DT imo.
- `unified-signatures`: Makes heavily overloaded method signatures harder to scan.
- `no-self-import`: Necessary so that some modules ("stream") can import shared types from `"bun"` without top-level imports. Relative imports (the recommended eslint fix) doesn't apply.
- `array-type`, `interface-over-type-literal`: I just really dislike these and don't think they represent a consensus view

I properly fixed the `ban-types` errors with per-line disablings though there are functions in `util/types` that use `String`, `Boolean`, etc for valid reasons.

Reply:
I understand `no-single-declare-module` for a runtime package, so no worries on that front.

For `no-single-element-tuple-type`, if there's a bug in the rule, I would suggest filing a bug on https://github.com/microsoft/DefinitelyTyped-tools.

The case you've listed for `no-unnecessary-generics` is the exact case where you wouldn't want to disable it; that is an unsafe API. If you must write that, disabling the lint at the specific API would be preferable so that the same mistake isn't made in other places. People really do accidentally introduce type variables they forget to use, or introduce them without realizing that they are not having any effect except to slow the compiler.

`redundant-undefined` [only applies to parameters](https://github.com/microsoft/DefinitelyTyped-tools/blob/b5c893dd7529da9cccc2e71451e74847f2703899/packages/eslint-plugin/src/rules/redundant-undefined.ts#L12), where `| undefined` is implied, and parameter properties are not (yet) legal `d.ts` syntax, so that should remain enabled on DT. 

For `unified-signatures`, I'd love to see an example. This is also one that is typically disabled inline when the rule is wrong; I haven't seen it disabled for style reasons.

For `no-self-import`, that makes sense for `declare module` packages, yes.

For `array-type` and `interface-over-type-literal`, I can understand those points of view, but DT is fairly consistent about this one...

Reply:
> I understand no-single-declare-module for a runtime package, so no worries on that front.

👍

> For no-single-element-tuple-type, if there's a bug in the rule, I would suggest filing a bug on https://github.com/microsoft/DefinitelyTyped-tools.

👍

> The case you've listed for no-unnecessary-generics is the exact case where you wouldn't want to disable it; that is an unsafe API. If you must write that, disabling the lint at the specific API would be preferable so that the same mistake isn't made in other places. People really do accidentally introduce type variables they forget to use, or introduce them without realizing that they are not having any effect except to slow the compiler.

I removed a couple actually unnecessary generics and did per-line exclusions everywhere else. 

> redundant-undefined [only applies to parameters](https://github.com/microsoft/DefinitelyTyped-tools/blob/b5c893dd7529da9cccc2e71451e74847f2703899/packages/eslint-plugin/src/rules/redundant-undefined.ts#L12), where | undefined is implied, and parameter properties are not (yet) legal d.ts syntax, so that should remain enabled on DT.

I see, sorry. I was glancing through our lint failures and thought I saw these in object types inside our params.

> For unified-signatures, I'd love to see an example. This is also one that is typically disabled inline when the rule is wrong; I haven't seen it disabled for style reasons.

Check out `stream.d.ts` L430 or so. Something like this. Keeping the handler for each event type on one line is easy to scan and update later if, say, the handler types for "pause" and "resume" diverge (despite the fact they're currently identical).

```ts
class Readable {
            addListener(event: "close", listener: () => void): this;
            addListener(event: "data", listener: (chunk: any) => void): this;
            addListener(event: "end", listener: () => void): this;
            addListener(event: "error", listener: (err: Error) => void): this;
            addListener(event: "pause", listener: () => void): this;
            addListener(event: "readable", listener: () => void): this;
            addListener(event: "resume", listener: () => void): this;
}
```

> For no-self-import, that makes sense for declare module packages, yes.

👍

> For array-type and interface-over-type-literal, I can understand those points of view, but DT is fairly consistent about this one...

Fixed all violations of `interface-over-type-literal`. 

There are a lot of false positives on `array-type`, basic repro:

```ts
interface Something {
  addListener(event: string | symbol, listener: (...args: any[]) => void): this;
}
```